pkg/crc/tls/tls.go

```diff
@@ -0,0 +1,277 @@
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to create certificate request")
+	}
+
+	csr, err := x509.ParseCertificateRequest(csrBytes)
+	if err != nil {
+		logging.Debugf("Failed to parse x509 certificate request: %s", err)
+		return nil, nil, errors.Wrap(err, "error parsing x509 certificate request")
+	}
+
+	// create a cert
+	cert, err := SignedCertificate(cfg, csr, key, caCert, caKey)
+	if err != nil {
+		logging.Debugf("Failed to create a signed certificate: %s", err)
+		return nil, nil, errors.Wrap(err, "failed to create a signed certificate")
+	}
+	return key, cert, nil
+}
+
+// GenerateSelfSignedCertificate generates a key/cert pair defined by CertCfg.
+func GenerateSelfSignedCertificate(cfg *CertCfg) (*rsa.PrivateKey, *x509.Certificate, error) {
+	key, err := PrivateKey()
+	if err != nil {
+		logging.Debugf("Failed to generate a private key: %s", err)
+		return nil, nil, errors.Wrap(err, "failed to generate private key")
+	}
+
+	crt, err := SelfSignedCertificate(cfg, key)
+	if err != nil {
+		logging.Debugf("Failed to create self-signed certificate: %s", err)
+		return nil, nil, errors.Wrap(err, "failed to create self-signed certificate")
+	}
+	return key, crt, nil
+}
+
+func GetRootCAKeyAndCert() (*rsa.PrivateKey, *x509.Certificate, error) {
+	rootCAConf := &CertCfg{
+		Subject:   pkix.Name{CommonName: "admin-kubeconfig-signer-custom", OrganizationalUnit: []string{"openshift"}},
+		KeyUsages: x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		Validity:  ValidityTenYears,
+		IsCA:      true,
+	}
+	return GenerateSelfSignedCertificate(rootCAConf)
+}
+
+func GetClientKeyAndCert(rootCAKey *rsa.PrivateKey, rootCACert *x509.Certificate) ([]byte, []byte, error) {
+	adminUserConf := &CertCfg{
+		Subject:      pkix.Name{CommonName: "system:admin", OrganizationalUnit: []string{"system:masters"}},
+		KeyUsages:    x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		ExtKeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+		Validity:     ValidityTenYears,
+	}
+	clientKey, clientCrt, err := GenerateSignedCertificate(rootCAKey, rootCACert, adminUserConf)
+	if err != nil {
+		return nil, nil, err
+	}
+	return PrivateKeyToPem(clientKey), CertToPem(clientCrt), nil
+}
+
+// PrivateKeyToPem converts an rsa.PrivateKey object to pem string
+func PrivateKeyToPem(key *rsa.PrivateKey) []byte {
+	keyInBytes := x509.MarshalPKCS1PrivateKey(key)
+	keyinPem := pem.EncodeToMemory(
+		&pem.Block{
+			Type:  "RSA PRIVATE KEY",
+			Bytes: keyInBytes,
+		},
+	)
+	return keyinPem
+}
+
+// CertToPem converts an x509.Certificate object to a pem string
+func CertToPem(cert *x509.Certificate) []byte {
+	certInPem := pem.EncodeToMemory(
+		&pem.Block{
+			Type:  "CERTIFICATE",
+			Bytes: cert.Raw,
+		},
+	)
+	return certInPem
+}
+
+func VerifyCertificateAgainstRootCA(rootPEM, clientPEM string) (bool, error) {
```

Comment:
This function takes a CA and a certificate (its name says so), but this is not reflected in the argument names? Maybe `CA` and `cert`, and leave the fact that they must be PEM-encoded up to a comment?